File: services/agent-engine/app/schemas/openclaw.py
"""Schemas for the OpenClaw control endpoints."""

from __future__ import annotations

from typing import Literal, Optional

from pydantic import BaseModel


class ToolPresence(BaseModel):
    installed: bool
    version: Optional[str] = None


class OpenClawConfigSummary(BaseModel):
    exists: bool
    path: str
    provider: Optional[str] = None
    model: Optional[str] = None


class GatewayInfo(BaseModel):
    running: bool
    port: int


class WebSearchInfo(BaseModel):
    provider: Optional[str] = None
    enabled: bool = False


class OpenClawStatusResponse(BaseModel):
    node: ToolPresence
    npm: ToolPresence
    ollama: ToolPresence
    openclaw: ToolPresence
    config: OpenClawConfigSummary
    gateway: GatewayInfo
    web_search: WebSearchInfo = WebSearchInfo()
    ready: bool


class WebSearchResult(BaseModel):
    ok: bool
    message: str = ""
    web_search: WebSearchInfo = WebSearchInfo()


class OpenClawInstallStatus(BaseModel):
    state: str  # idle | running | success | error
    message: str = ""
    log_tail: str = ""


class OpenClawConfigRequest(BaseModel):
    mode: Literal["free", "api"] = "free"
    model: str = "gemma4:latest"
    provider: Optional[str] = None  # for mode="api" (e.g. anthropic, openai)
    api_key: Optional[str] = None
    base_url: Optional[str] = None  # for mode="free" (Ollama URL)


class OpenClawConfigResult(BaseModel):
    ok: bool
    message: str = ""
    config: OpenClawConfigSummary


class GatewayActionResult(BaseModel):
    ok: bool
    message: str = ""
    running: bool = False


class DashboardUrlResponse(BaseModel):
    ok: bool
    url: Optional[str] = None
    message: str = ""


class ChannelInfo(BaseModel):
    slug: str
    name: str
    icon: str = "💬"
    connect: str = "token"  # token | login
    can_add: bool = True
    installed: bool = False
    configured: bool = False
    popular: bool = False


class ChannelsResponse(BaseModel):
    ok: bool
    message: str = ""
    channels: list[ChannelInfo] = []


class AddChannelRequest(BaseModel):
    channel: str
    token: str


class ChannelActionResult(BaseModel):
    ok: bool
    message: str = ""


class OpenClawAgent(BaseModel):
    id: str
    name: str = ""
    model: str = ""
    workspace: str = ""
    is_default: bool = False
    bindings: int = 0


class OpenClawAgentsResponse(BaseModel):
    ok: bool
    message: str = ""
    agents: list[OpenClawAgent] = []


class CreateOpenClawAgentRequest(BaseModel):
    name: str
    model: Optional[str] = None  # defaults to ollama/gemma4:latest (free)
    instructions: Optional[str] = None
    emoji: Optional[str] = None


class OpenClawAgentResult(BaseModel):
    ok: bool
    message: str = ""
    agent: Optional[OpenClawAgent] = None


class AgentActionResult(BaseModel):
    ok: bool
    message: str = ""


class FileHandoffRequest(BaseModel):
    from_agent_id: str
    to_agent_id: str
    file_name: str


class FileHandoffResult(BaseModel):
    ok: bool
    message: str = ""
    path: Optional[str] = None


class AgentChatRequest(BaseModel):
    message: str
    session_id: Optional[str] = None  # stable id → conversation continuity
    activity_task: Optional[str] = None  # short label shown in the Office view
    activity_kind: Optional[str] = None  # chat | team (anything else falls back to chat)


class AgentChatResult(BaseModel):
    ok: bool
    reply: str = ""
    model: str = ""
    session_id: str = ""
    message: str = ""


class AgentWorkFile(BaseModel):
    name: str
    path: str  # relative to the agent's workspace
    size_bytes: int = 0
    modified_at: int = 0  # epoch ms
    abs_path: str = ""


class AgentDocumentsGroup(BaseModel):
    agent_id: str
    name: str
    emoji: str = ""
    files: list[AgentWorkFile] = []


class AgentDocumentsResponse(BaseModel):
    ok: bool
    message: str = ""
    agents: list[AgentDocumentsGroup] = []


class AgentDocumentContent(BaseModel):
    ok: bool
    message: str = ""
    name: str = ""
    content: str = ""
    truncated: bool = False


class AgentDiscordStatus(BaseModel):
    ok: bool
    connected: bool
    account_id: str = ""
    gateway_running: bool = False
    message: str = ""


class ConnectDiscordRequest(BaseModel):
    bot_token: str


class AgentDiscordResult(BaseModel):
    ok: bool
    message: str = ""


class AgentFile(BaseModel):
    name: str
    label: str
    size: int = 0
    content: str = ""
    exists: bool = False


class AgentFilesResponse(BaseModel):
    ok: bool
    message: str = ""
    workspace: str = ""
    files: list[AgentFile] = []


class SaveAgentFileRequest(BaseModel):
    content: str = ""


class SessionSummary(BaseModel):
    session_id: str
    preview: str = ""
    message_count: int = 0
    updated_at: int = 0  # epoch ms
    size_bytes: int = 0
    # Where the chat came from: evano | team | discord | telegram | whatsapp |
    # slack | dashboard | subagent | cron | other
    origin: str = "other"


class AgentChats(BaseModel):
    agent_id: str
    name: str
    emoji: str = ""
    sessions: list[SessionSummary] = []


class AllChatsResponse(BaseModel):
    ok: bool
    message: str = ""
    agents: list[AgentChats] = []


class SessionsResponse(BaseModel):
    ok: bool
    message: str = ""
    sessions: list[SessionSummary] = []
    total_bytes: int = 0


class SessionMessage(BaseModel):
    role: str
    content: str


class SessionDetail(BaseModel):
    ok: bool
    message: str = ""
    messages: list[SessionMessage] = []


class ClearSessionsRequest(BaseModel):
    older_than_days: Optional[int] = None  # None/0 = clear all


class ClearSessionsResult(BaseModel):
    ok: bool
    message: str = ""
    deleted: int = 0
    freed_bytes: int = 0

File: apps/desktop/src/renderer/src/lib/api/types/openclaw.ts
/**
 * OpenClaw control: install, config, gateway, channels, agents, sessions, files, web search.
 *
 * Mirrors the FastAPI schemas in `services/agent-engine/app/schemas`.
 */

// ---- OpenClaw control -----------------------------------------------------

export interface ToolPresence {
  installed: boolean;
  version: string | null;
}

export interface OpenClawConfigSummary {
  exists: boolean;
  path: string;
  provider: string | null;
  model: string | null;
}

export interface GatewayInfo {
  running: boolean;
  port: number;
}

export interface WebSearchInfo {
  provider: string | null;
  enabled: boolean;
}

/** Response from `GET /openclaw/status`. */
export interface OpenClawStatus {
  node: ToolPresence;
  npm: ToolPresence;
  ollama: ToolPresence;
  openclaw: ToolPresence;
  config: OpenClawConfigSummary;
  gateway: GatewayInfo;
  web_search: WebSearchInfo;
  ready: boolean;
}

/** Response from `POST /openclaw/web-search/enable`. */
export interface WebSearchResult {
  ok: boolean;
  message: string;
  web_search: WebSearchInfo;
}

/** Response from `POST /openclaw/install` and `GET /openclaw/install/status`. */
export interface OpenClawInstallStatus {
  state: string; // idle | running | success | error
  message: string;
  log_tail: string;
}

/** A prerequisite Evano can install for you (download + launch official installer). */
export type PrereqTarget = "node" | "ollama";

/** Response from `/openclaw/prereqs/{target}/install[/status]`. */
export interface PrereqInstallStatus {
  state: string; // idle | downloading | launching | launched | error
  message: string;
  percent: number;
  download_url: string; // official page, manual fallback
}

/** Response from `POST /openclaw/config`. */
export interface OpenClawConfigResult {
  ok: boolean;
  message: string;
  config: OpenClawConfigSummary;
}

/** Response from `POST /openclaw/gateway/{start,stop}`. */
export interface GatewayActionResult {
  ok: boolean;
  message: string;
  running: boolean;
}

/** Response from `GET /openclaw/dashboard`. */
export interface DashboardUrlResponse {
  ok: boolean;
  url: string | null;
  message: string;
}


/** A messaging platform OpenClaw can connect to (WhatsApp, Discord, …). */
export interface OpenClawChannel {
  slug: string;
  name: string;
  icon: string;
  connect: "token" | "login";
  can_add: boolean;
  installed: boolean;
  configured: boolean;
  popular: boolean;
}

/** Response from `GET /openclaw/channels`. */
export interface ChannelsResponse {
  ok: boolean;
  message: string;
  channels: OpenClawChannel[];
}

/** Response from `POST /openclaw/channels/add`. */
export interface ChannelActionResult {
  ok: boolean;
  message: string;
}

/** An OpenClaw agent (runs inside OpenClaw, with its own workspace + model). */
export interface OpenClawAgent {
  id: string;
  name: string;
  model: string;
  workspace: string;
  is_default: boolean;
  bindings: number;
}

/** Response from `GET /openclaw/agents`. */
export interface OpenClawAgentsResponse {
  ok: boolean;
  message: string;
  agents: OpenClawAgent[];
}

/** Response from `POST /openclaw/agents`. */
export interface OpenClawAgentResult {
  ok: boolean;
  message: string;
  agent: OpenClawAgent | null;
}

/** Response from `DELETE /openclaw/agents/{id}` and file saves. */
export interface AgentActionResult {
  ok: boolean;
  message: string;
}

/** Response from `POST /openclaw/file-handoff` (copy a file agent→agent). */
export interface FileHandoffResult {
  ok: boolean;
  message: string;
  path?: string | null;
}

/** Response from `POST /openclaw/agents/{id}/chat`. */
export interface AgentChatResult {
  ok: boolean;
  reply: string;
  model: string;
  session_id: string;
  message: string;
}

/** A work file an agent created in its workspace. */
export interface AgentWorkFile {
  name: string;
  path: string; // relative to the agent's workspace
  size_bytes: number;
  modified_at: number; // epoch ms
  abs_path: string;
}

/** One agent's work files (the Documents tab). */
export interface AgentDocumentsGroup {
  agent_id: string;
  name: string;
  emoji: string;
  files: AgentWorkFile[];
}

/** Response from `GET /openclaw/documents`. */
export interface AgentDocumentsResponse {
  ok: boolean;
  message: string;
  agents: AgentDocumentsGroup[];
}

/** Response from `GET /openclaw/documents/content`. */
export interface AgentDocumentContent {
  ok: boolean;
  message: string;
  name: string;
  content: string;
  truncated: boolean;
}

/** Response from `GET /openclaw/agents/{id}/discord`. */
export interface AgentDiscordStatus {
  ok: boolean;
  connected: boolean;
  account_id: string;
  gateway_running: boolean;
  message: string;
}

/** Response from `POST/DELETE /openclaw/agents/{id}/discord`. */
export interface AgentDiscordResult {
  ok: boolean;
  message: string;
}

/** One config file in an OpenClaw agent's workspace. */
export interface AgentFile {
  name: string;
  label: string;
  size: number;
  content: string;
  exists: boolean;
}

/** Response from `GET /openclaw/agents/{id}/files`. */
export interface AgentFilesResponse {
  ok: boolean;
  message: string;
  workspace: string;
  files: AgentFile[];
}

/** Where a conversation came from. */
export type ChatOrigin =
  | "evano"
  | "team"
  | "discord"
  | "telegram"
  | "whatsapp"
  | "slack"
  | "signal"
  | "imessage"
  | "dashboard"
  | "subagent"
  | "cron"
  | "other";

/** A saved conversation with an agent. */
export interface SessionSummary {
  session_id: string;
  preview: string;
  message_count: number;
  updated_at: number; // epoch ms
  size_bytes: number;
  origin: ChatOrigin;
}

/** One agent's conversations (the Chats tab). */
export interface AgentChats {
  agent_id: string;
  name: string;
  emoji: string;
  sessions: SessionSummary[];
}

/** Response from `GET /openclaw/chats`. */
export interface AllChatsResponse {
  ok: boolean;
  message: string;
  agents: AgentChats[];
}

/** Response from `GET /openclaw/agents/{id}/sessions`. */
export interface SessionsResponse {
  ok: boolean;
  message: string;
  sessions: SessionSummary[];
  total_bytes: number;
}

/** Response from `GET /openclaw/agents/{id}/sessions/{sid}`. */
export interface SessionDetail {
  ok: boolean;
  message: string;
  messages: { role: "user" | "assistant"; content: string }[];
}

/** Response from `POST /openclaw/agents/{id}/sessions/clear`. */
export interface ClearSessionsResult {
  ok: boolean;
  message: string;
  deleted: number;
  freed_bytes: number;
}


/** A customer-facing channel offered in the Customer Service setup. */
export interface SupportChannel {
  slug: string;
  name: string;
  icon: string;
  connect: "token" | "login";
  configured: boolean;
}

/** A live routing: this channel's messages go to this agent. */
export interface SupportAssignment {
  agent_id: string;
  channel: string;
  account_id: string;
}

export interface SupportAgentLite {
  id: string;
  name: string;
  emoji: string;
}

/** Response from `GET /openclaw/support/status`. */
export interface CustomerServiceStatus {
  ok: boolean;
  message: string;
  gateway_running: boolean;
  agents: SupportAgentLite[];
  channels: SupportChannel[];
  assignments: SupportAssignment[];
}

/** Response from the in-app WhatsApp pairing endpoints. */
export interface WhatsAppLoginStatus {
  state: "idle" | "starting" | "qr" | "connected" | "error" | "expired";
  qr_svg: string | null;
  message: string;
}
